### PR TITLE
Apply refactoring

### PR DESCRIPTION
**Pull request by metanet**
* Convert workspace paths (i.e., paths given by the user under --dir) to
absolute paths via _to_real_path() and _get_absolute_path() functions in
order to unify path manipulations and security checks.
* Transform absolute paths to workspace paths via _get_workspace_path()
to eliminate duplicate code pieces.
* Unify failure responses via _failure_json().
* Enable file creation only with allowed extensions.
* Remove '\00' (NULL terminator) from ag and ack outputs.
* Apply minor renamings on some functions and local vars.
* Return file modification time (to be used for conflict detection on
save).

### CODE REVIEW

File: pervane/serve.py [hunk 1]
```diff
@@ -304,62 +304,59 @@ def make_tree(path):
...
-def _get_workspace_path(fn):
-  workspace_path = fn.replace(_get_root_dir(), '')
-  # Prepend separator in case user enter root dir with / at the
-  # end.
-  if (not workspace_path.startswith(os.sep)):
-    workspace_path = os.sep + workspace_path
-  return workspace_path
+def _get_workspace_path(path):
+  path = path.replace(_get_root_dir(trailing_separator=False), '')
+  return path if path.startswith(os.sep) else os.sep + path
```
Comment (hakanu) [inline pervane/serve.py:309]: os.path.join(os.sep, path) instead of `os.sep + path`

File: pervane/serve.py [hunk 2]
```diff
@@ -371,6 +368,80 @@ def _get_file_mode(path):
   return 'text/html'
 
 
+def _get_file_mod_time(path):
+  mod_secs = os.path.getmtime(path)
```
Comment (hakanu) [inline pervane/serve.py:372]: you can combine these two lines.

File: pervane/serve.py [hunk 3]
```diff
@@ -371,6 +368,80 @@ def _get_file_mode(path):
...
+  return request.args.get(arg_name, '').strip()
+
+
+def _get_request_json(arg_name):
+  return request.json.get(arg_name, '').strip()
+
+
+def _to_real_path(parent_path, relative_path):
+  # ignores the heading '/' in relative_path
+  start = 1 if relative_path.startswith('/') else 0
```
Comment (hakanu) [inline pervane/serve.py:386]: nit: `start_index` sounds better.

File: pervane/serve.py [hunk 4]
```diff
@@ -371,6 +368,80 @@ def _get_file_mode(path):
...
+  root_dir = _get_root_dir(trailing_separator=False)
+  path = _to_real_path(root_dir, path)
+
+  if path.startswith(root_dir):
+    return (path, None)
+
+  return (None, 'Not authorized')
+
+
+def _get_new_node_name(new_node_name):
```
Comment (hakanu) [inline pervane/serve.py:404]: can you add  a docstring about the return type. Returning tuple and it's not obvious which element is what. Probably we should use namedtuple or an object. But you can ignore for that for this PR.

File: pervane/serve.py [hunk 5]
```diff
@@ -371,6 +368,80 @@ def _get_file_mode(path):
...
+  new_node_name = new_node_name.strip()
+
+  if new_node_name.startswith(os.path.sep):
+    new_node_name = new_node_name[1:].strip()
+
+  if not new_node_name:
+    return (None, None, "empty file name!")
+
+  if new_node_name.endswith(os.path.sep):
+    # eliminate the tailing file separator
```
Comment (hakanu) [inline pervane/serve.py:417]: nit: capitalize the first letter and dot at the end in the comment line.

File: pervane/serve.py [hunk 6]
```diff
@@ -371,6 +368,80 @@ def _get_file_mode(path):
...
+  if new_node_name == '.' or new_node_name == '..':
+    return (None, None, "invalid file name: " + new_node_name)
+
+  if '.' in new_node_name:
+    file_name, extension = os.path.splitext(new_node_name)
+    if extension not in args.note_extensions:
+      return (None, None, "invalid note extension: " + new_node_name)
+  else:
+    # append .md if there is no extension
+    file_name = new_node_name + ".md"
```
Comment (hakanu) [inline pervane/serve.py:433]: we use single quote strings throughout the code. Let's be consistent here.

File: pervane/serve.py [hunk 7]
```diff
@@ -371,6 +368,80 @@ def _get_file_mode(path):
...
+
+    return (dir_name, None, None)
+
+  if new_node_name == '.' or new_node_name == '..':
+    return (None, None, "invalid file name: " + new_node_name)
+
+  if '.' in new_node_name:
+    file_name, extension = os.path.splitext(new_node_name)
+    if extension not in args.note_extensions:
+      return (None, None, "invalid note extension: " + new_node_name)
```
Comment (hakanu) [inline pervane/serve.py:430]: single quote strings for consistency

File: pervane/serve.py [hunk 8]
```diff
@@ -400,55 +471,44 @@ def front_page_handler():
...
+
+    return jsonify({
       'result': 'success',
       'content': html_content,
-  })
+      'mod_time': mod_time
+    })
+  except Exception as e:
+    logging.error('There is an error while reading: %s', str(e))
+    # don't leak verified path
```
Comment (hakanu) [inline pervane/serve.py:510]: nit: Capitalize first letter and . at the end.

**Comment by metanet** [comment]
updated the PR based on your review. please have another look @hakanu

**Pull request merged**